File: apps/driver-android/src/main/kotlin/lk/mageride/driver/menu/MenuScreen.kt
package lk.mageride.driver.menu

import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalDrawerSheet
import androidx.compose.material3.NavigationDrawerItem
import androidx.compose.material3.NavigationDrawerItemDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.nav.DriverRoute
import lk.mageride.driver.ui.component.DriverHeader
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import org.koin.androidx.compose.koinViewModel

/**
 * **SCR-DA-036 · the menu drawer.**
 *
 * The wireframe draws a Material 3 **modal navigation drawer** with a `primaryContainer` header —
 * avatar, name, level badge, driver id and rating — over the dashboard, and eight rows under it.
 *
 * **AL-31 is why it is a tab rather than a corner affordance.** *"The dashboard has NO top-left
 * hamburger; navigation is the bottom-nav Menu tab"*, and `DriverTab.MenuTab` is a peer of Home. So
 * this destination **is** the drawer: the sheet is drawn at its documented width against a scrim,
 * and the scrim takes the driver back where they came from.
 *
 * **Δ MCS-24 — both halves of "scrim tap / swipe-left closes" are wired now.** Only the tap was. On
 * a sheet pinned to the left edge, dragging it leftward is the gesture a driver reaches for first,
 * and it did nothing at all — no movement, no rubber-band, no hint that the tap was the way out.
 *
 * **Δ MCS-24 — the header is the DRIVER, not the app.** It printed `app_name`, and the reason was
 * recorded and was a good one: this component did not own the profile read, so *"a wrong name is
 * worse than none, and inventing a rating would be worse still"*. That reasoning survives — the
 * rating is still drawn only if one exists, and none does — but the premise does not.
 * [MenuViewModel] owns the read now, so the name, the level and the live vehicle's plate are this
 * screen's to show. What the header must never do is guess, and it does not.
 *
 * @param onClose The scrim tap, and now the swipe. Goes back rather than to Home: a driver who
 *   opened Menu from the Wallet tab expects to return to the Wallet tab.
 */
@Composable
internal fun MenuScreen(onOpen: (DriverRoute) -> Unit, onClose: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: MenuViewModel = koinViewModel()
    val header by viewModel.header.collectAsStateWithLifecycle()

    Row(modifier = modifier.fillMaxSize()) {
        ModalDrawerSheet(
            modifier = Modifier
                .width(ControlTokens.DrawerWidth)
                // The whole gesture is accumulated and judged on RELEASE rather than acted on per
                // event, so a drag that wanders left and comes back does not close: a driver who
                // starts a swipe and changes their mind keeps their menu.
                .pointerInput(Unit) {
                    var travelled = 0f

                    detectHorizontalDragGestures(
                        onDragStart = { travelled = 0f },
                        onDragCancel = { travelled = 0f },
                        onDragEnd = {
                            if (travelled <= -CLOSE_SWIPE_FRACTION * size.width) onClose()
                            travelled = 0f
                        },
                        onHorizontalDrag = { change, dragAmount ->
                            change.consume()
                            travelled += dragAmount
                        },
                    )
                },
        ) {
            DriverHeader(
                state = header,
                modifier = Modifier.padding(
                    horizontal = MageRideTheme.spacing.md,
                    vertical = MageRideTheme.spacing.sm,
                ),
            )

            Column(
                modifier = Modifier.verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
            ) {
                MenuDestination.entries.forEach { destination ->
                    NavigationDrawerItem(
                        label = { Text(text = stringResource(destination.label)) },
                        icon = { Icon(imageVector = destination.icon, contentDescription = null) },
                        selected = false,
                        onClick = { onOpen(destination.route) },
                        modifier = Modifier.padding(NavigationDrawerItemDefaults.ItemPadding),
                    )
                }
            }
        }

        // The wireframe's scrim over the dashboard. Tapping it is the documented way out.
        Surface(
            modifier = Modifier
                .weight(1f)
                .fillMaxSize(),
            color = MaterialTheme.colorScheme.scrim.copy(alpha = SCRIM_ALPHA),
            onClick = onClose,
            content = {},
        )
    }
}

/**
 * How far left the sheet must be dragged before releasing closes it (Δ MCS-24).
 *
 * A fraction of the sheet rather than a fixed distance: the drawer is [ControlTokens.DrawerWidth]
 * wide on every handset, so a third of it is a third of the same control everywhere. Low enough to
 * be a flick, high enough that a horizontal wobble during a vertical scroll of the rows below does
 * not dismiss the screen.
 */
private const val CLOSE_SWIPE_FRACTION = 0.33f

/** D2' §0.2's scrim opacity — the same figure the offline map overlay uses. */
private const val SCRIM_ALPHA = 0.45f

File: apps/driver-ios/DriverApp/Menu/MenuScreen.swift
import SwiftUI

/// **SCR-DI-036 · the Menu tab.**
///
/// The wireframe draws a large title, a driver card and one grouped list of eight rows over a tab bar
/// with **Menu** selected.
///
/// **AL-31 is why this is a tab rather than a corner affordance.** *"The dashboard has NO top-left
/// hamburger; navigation is the bottom Menu tab"*, and ``DriverTab/menu`` is a peer of Home.
///
/// **Δ iOS — this is a `List`, where the Android twin is a `ModalDrawerSheet`.** The cell's own `Δ iOS`
/// clause is *"`List` + `NavigationLink`"*, and it is the right shape here for a reason beyond the
/// wireframe: a modal drawer is Material's answer to a navigation surface, and on this platform a tab
/// whose content is a scrim with a panel beside it would be a drawer wearing a tab's clothes. The rows
/// push onto the Menu tab's own stack, which is what makes the system's `‹ Menu` back button say the
/// right thing on every screen that hangs off here.
///
/// The rows are `Button`s rather than `NavigationLink`s because ``DriverNavigator`` owns the four
/// stacks — a `NavigationLink(value:)` would push onto whichever stack it happened to be inside and
/// leave the navigator's copy of the path behind. The destination is still ``DriverRoute``'s, and it is
/// still resolved by the one `navigationDestination` in the app.
@MainActor
struct MenuScreen: View {

    @StateObject private var model: MenuModel
    private let onOpen: (DriverRoute) -> Void

    /// `@autoclosure` and `_model =`, the same shape SCR-DI-029 uses: a `@StateObject` must be
    /// assigned through its wrapper, and building the model eagerly at the call site would
    /// construct one on every parent redraw and throw it away.
    init(model: @autoclosure @escaping () -> MenuModel, onOpen: @escaping (DriverRoute) -> Void) {
        _model = StateObject(wrappedValue: model())
        self.onOpen = onOpen
    }

    var body: some View {
        List {
            Section {
                driverCard
                    .listRowInsets(EdgeInsets())
                    .listRowBackground(Color.clear)
            }

            Section {
                ForEach(MenuDestination.allCases) { destination in
                    row(destination)
                }
            }
        }
        .task {
            // Δ MCS-27 — the cache first, so the tab opens on a name; the reads refresh behind it.
            await model.paintFromCache()
            await model.load()
        }
        .task { await model.loadPhoto() }
        .listStyle(.insetGrouped)
        .navigationTitle(Text(key: "nav_menu"))
        .background(MageRideColor.background)
    }

    /// The wireframe's driver card.
    ///
    /// **Δ MCS-24 — the DRIVER, not a label.** It drew `menu_driver` above the platform id, and the
    /// reason was recorded and was a good one: this component did not own the profile read, so *"a
    /// wrong name is worse than none and an invented rating is worse still"*. That reasoning
    /// survives — the rating is drawn only if one exists, and none does — but the premise does not.
    /// ``MenuModel`` owns the read now.
    ///
    /// The layout is ``DriverHeader``, which SCR-DI-029 also draws. Before this they were two
    /// independent pieces of layout for one block, and they had drifted in the same direction: both
    /// showed an identifier where the wireframe draws the vehicle.
    private var driverCard: some View {
        DriverHeader(state: model.header)
            .padding(MageRideSpacing.sm)
    }

    private func row(_ destination: MenuDestination) -> some View {
        Button { onOpen(destination.route) } label: {
            HStack(spacing: MageRideSpacing.sm) {
                Image(systemName: destination.symbolName)
                    .font(.footnote)
                    .foregroundStyle(MageRideColor.onStatus)
                    .frame(width: MageRideControl.listRowIcon, height: MageRideControl.listRowIcon)
                    .background(
                        destination.accent,
                        in: RoundedRectangle(cornerRadius: MageRideRadius.sm, style: .continuous)
                    )

                Text(key: destination.labelKey)
                    .mageFont(.body)
                    .foregroundStyle(MageRideColor.onSurface)

                Spacer(minLength: MageRideSpacing.xs)

                Image(systemName: "chevron.right")
                    .font(.footnote)
                    .foregroundStyle(MageRideColor.outlineVariant)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityAddTraits(.isButton)
    }
}
